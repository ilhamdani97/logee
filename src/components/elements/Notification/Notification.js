/* eslint-disable react/no-danger */
import React, { useCallback, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';
import moment from 'moment';
import { generateTextNotif, formatFullDate } from '../../../utils/format';


function Notification(props) {
  const history = useHistory();
  const { className, show, handleShow, showAll } = props;
  const { fetchData, listNotifikasi } = useSelector(s => s.listNotifikasi);
  const dispatch = useDispatch();
  const emptyList = !listNotifikasi || listNotifikasi.length === 0;

  const classes = [styles.root, className].filter(Boolean).join(' ');

  useEffect(() => {
    dispatch(fetchData('listNotifikasi'));
  }, [dispatch, show]);

  const _handleGoto = useCallback((item) => {
    if (item.type === 'order-created' || item.type === 'order-finished') {
      const _deliveryDetailId = item.deliveryDetailId;
      if (_deliveryDetailId) {
        history.push(`/pemesanan?id=${_deliveryDetailId}`);
      }
    } else if (item.type === 'vehicle') {
      history.push(`/kendaraan?id=${item.vehicleId}`);
    }
  }, []);

  return (
    <div className={[classes, show ? styles.showNotifAlert : ''].join(' ')} onMouseLeave={() => handleShow(false)} >
      <div className={styles.boxcursor} />
      <div className={styles.containerNotif}>
        <section>
          <div>
            Notifikasi
          </div>
          <div onClick={() => showAll()}>
            Lihat Semua
          </div>
        </section>
        <section className={emptyList ? styles.emptyList : ''}>
          <div>
            {emptyList ?
              <div className={styles.listNotifEmpty}>Belum ada notifikasi</div>
              :
              listNotifikasi.map((item, i) => {
                const _textNotif = generateTextNotif(item.text);

                return (
                  <div className={styles.listNotif} key={i} onClick={() => _handleGoto(item, i)} >
                    <div className={i !== 0 ? styles.hideCreateAt : ''}>
                      {formatFullDate(item.createdAt, 4)}
                    </div>
                    <div>
                      <div className={item.isClicked ? styles.hasRead : ''} />
                      <div>
                        <span dangerouslySetInnerHTML={{ __html: _textNotif.text1 || '' }} />
                        <span>{_textNotif.text2}</span>
                        <span dangerouslySetInnerHTML={{ __html: _textNotif.text3 || '' }} />
                      </div>
                    </div>
                    <div>
                      {moment(item.createdAt).format('HH:mm')}
                    </div>
                  </div>
                );
              })
            }
          </div>
        </section>
      </div>
    </div>
  );
}

export default React.memo(Notification, (prevProps, nextProps) => {
  return prevProps.show === nextProps.show;
});

Notification.defaultProps = {
  className: '',
  handleShow: () => { },
  show: false,
  showAll: () => { },
};

Notification.propTypes = {
  className: PropTypes.string,
  handleShow: PropTypes.func,
  show: PropTypes.bool,
  showAll: PropTypes.func
};
