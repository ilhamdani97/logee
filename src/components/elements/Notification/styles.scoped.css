.root {
  position: fixed;
  left: 2.2rem;
  background-color: transparent;
  bottom: 2rem;
  padding: 0.6375rem;
  z-index: 999;
  display: none;
}

.showNotifAlert{
  display: block !important;
}

.boxcursor{
  position: fixed;
  left: 2.5rem;
  z-index: 1;
  transform: rotate(45deg);
  border-bottom-left-radius: 2.5px;
  background-color: var(--background-color-2);
  width: 1rem;
  height: 1rem;
  bottom:4.3rem;
}
.containerNotif{
  position: relative;
  z-index: 777;
  background-color: var(--background-color-2);
  padding: 1rem 0;
  border-radius: 0.2rem;
  box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.12);
  border-radius: 5px;
  min-width: 18rem;
  max-width:18rem;
  min-height: 25.5rem;
  max-height: 25.5rem;
  overflow: hidden;
  
}
.containerNotif>section:nth-of-type(1){
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 0 1rem 0.8rem;

}

.containerNotif>section:nth-of-type(1)>div:nth-of-type(1){
  color: var(--text-color);
  font-size: 1.125rem;
  font-weight: 700;
}

.containerNotif>section:nth-of-type(1)>div:nth-of-type(2){
  font-size: 0.625rem;
  cursor: pointer;
  background-color: var(--grey-30);
  width: 4.125rem;
  height: 1.5rem;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 0.25rem;
}

.containerNotif>section:nth-of-type(1)>div:nth-of-type(2):hover{
  background-color: #F76707;
  color: #FFFFFF;
}

.containerNotif>section:nth-of-type(2){
  display: flex;
  flex-direction: column;
  max-height: 22rem;
  overflow-y: auto;
  padding-bottom: 5rem;
  margin-top: 0.5rem;
}

.emptyList{
  overflow-y: hidden !important;
}

.activeTabNotification{
  border-bottom: 2px solid var(--primary-color);
  color: var(--primary-color) !important;
  font-weight: 600;
}

.listNotif{
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: flex-start;
  cursor: pointer;
}

.listNotif>div:nth-of-type(1){
  font-size: 0.675rem;
  color:var(--grey-color-2);
  padding-left: 1rem;
  margin-bottom: 0.5rem;
}


.listNotif>div:nth-of-type(2){
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.listNotif>div:nth-of-type(3){
  font-size: 0.675rem;
  color:var(--grey-color-2);
  padding-left: 1.5rem;
  margin-bottom: 0.5rem;
}


.listNotif>div:nth-of-type(2)>div:nth-of-type(1){
 width: 0.3125rem;
 height: 0.3125rem;
 border-radius: 0.15rem;
 background-color: var(--error-color);
 margin-right: 0.35rem;
 margin-left: 1rem;
 margin-bottom: 0.5rem;

}

.listNotif>div:nth-of-type(2)>div:nth-of-type(2){
  width: 14.125rem;
  height: 2.125rem;
  word-wrap: break-word;
  white-space:wrap;
  word-break:break-all;
  font-size: 0.9rem;
  color:var(--black-4);
  margin-bottom: 0.5rem;
  margin-right: 0.5rem;
  text-overflow: ellipsis;
  overflow: hidden;
  display: -webkit-box;
    -webkit-line-clamp: 2;
    -webkit-box-orient: vertical;
 }

 .listNotif>div:nth-of-type(2)>div:nth-of-type(2)>span:nth-of-type(1){
  font-style: normal;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.1rem;
  color: #212121;
  word-break: break-word;
 }

 .listNotif>div:nth-of-type(2)>div:nth-of-type(2)>span:nth-of-type(2){
  font-size: 0.875rem;
  margin-bottom: 0.5rem;
  font-weight: bold;
  word-break: break-word;
 }

 .hideCreateAt{
   display: none;
 }
 
 .hasRead{  
  width: 0.5rem;
  height: 0.5rem;
  border-radius: 0.25rem;
  background-color: var(--grey-color-2) !important;
  margin-right: 0.625rem;
  margin-bottom: 0.5rem;
 }

.listNotifEmpty{
  padding: 0 1rem;
  font-size: 0.75rem;
  font-weight: normal;
  color: var(--grey-color-2);
 }

::-webkit-scrollbar {
  width: 0.3125rem;
}

::-webkit-scrollbar-thumb {
  background:#BDBDBD;
  border-radius: 0.375rem;
}




