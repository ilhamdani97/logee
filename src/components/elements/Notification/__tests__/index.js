import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Notification from '../Notification';
import { useSelector } from 'react-redux';

const { defaultProps } = Notification.type;

useSelector.mockImplementation(fn => {
  fn({});
  return {
    listNotifikasi: [

      {
        messageId: 'f092f8cf-8282-4bb3-87a2-10253911eabe',
        createdAt: '2021-08-03',
        deliveryDetailId: 'c8e84305-7900-432a-8de7-84ac96c9a6b9',
        text: 'Order baru layanan truk <span>No. Pesanan LT-202108030021</span> oleh <b>BUDI JAYA SEJAHTERA ABADI</b>menggunakan truk milik <b>BINTANG EXPRESS</b> dari <b>Kec. Cibeunying Kidul</b> menuju <b>Malang</b> nilai transaksi <b>Rp500.000,00</b>',
        type: 'order-created'
      },
      {
        messageId: 'f092f8cf-8282-4bb3-87a2-10253911eabe',
        createdAt: '2021-08-03',
        deliveryDetailId: 'c8e84305-7900-432a-8de7-84ac96c9a6b9',
        text: 'Order baru layanan truk <span>No. Pesanan LT-202108030021</span> oleh <b>BUDI JAYA SEJAHTERA ABADI</b>menggunakan truk milik <b>BINTANG EXPRESS</b> dari <b>Kec. Cibeunying Kidul</b> menuju <b>Malang</b> nilai transaksi <b>Rp500.000,00</b>',
        type: 'order-created'
      },
    ],
    fetchData: jest.fn(),
    meta: ''
  };
});

describe('src/components/elements/Notification', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Notification {...defaultProps} />);
    defaultProps.handleShow();
    expect(tree).toMatchSnapshot();
  });

  test('handle Close Notif', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Notification {...defaultProps} />);
    const WrapCloseNotif = tree.props.children[0].type;
    const _handleSetActiveTab = tree.props.onMouseLeave;
    _handleSetActiveTab(false);
    expect(WrapCloseNotif).toBe('div');
  });


  test('handle Goto Order Detail', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Notification {...defaultProps}  />);
    const WrapGotoOrderDetail = tree.props.children[1].props.children[1]
      .props.children.props.children[0].key;

    const _handleGoto = tree.props.children[1].props.children[1]
      .props.children.props.children[0].props.onClick;

    const textNotif = {
      createdAt: '2021-01-08',
      deliveryDetailId: '7914723e-799e-417b-aab7-cfeb5b98a7bf',
      isClicked: false,
      isDeleted: false,
      messageId: '8af7980b-465b-4450-affa-b48509b21bf4',
      text: 'Order baru layanan truck <span>LT-202101080055</span> dari <b>undefined</b> menggunakan truk milik â†µ  <b>PT AZZAM LOGISTICS</b> dari <b>undefined</b> menuju <b>{originToDestination}</b> nilai transaksi <b>Rpundefined</b>',
      type: 'order-created'
    };

    _handleGoto(textNotif);
    expect(WrapGotoOrderDetail).toBe('0');
  });
});
