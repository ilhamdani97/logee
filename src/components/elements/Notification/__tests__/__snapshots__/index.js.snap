// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Notification render 1`] = `
<div
  className=" "
  onMouseLeave={[Function]}
>
  <div />
  <div>
    <section>
      <div>
        Notifikasi
      </div>
      <div
        onClick={[Function]}
      >
        Lihat Semua
      </div>
    </section>
    <section
      className=""
    >
      <div>
        <div
          onClick={[Function]}
        >
          <div
            className=""
          >
            DD MMM YYYY
          </div>
          <div>
            <div
              className=""
            />
            <div>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Order baru layanan truk ",
                  }
                }
              />
              <span>
                No. Pesanan LT-202108030021
              </span>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": " oleh <b>BUDI JAYA SEJAHTERA ABADI</b>menggunakan truk milik <b>BINTANG EXPRESS</b> dari <b>Kec. Cibeunying Kidul</b> menuju <b>Malang</b> nilai transaksi <b>Rp500.000,00</b>",
                  }
                }
              />
            </div>
          </div>
          <div>
            HH:mm
          </div>
        </div>
        <div
          onClick={[Function]}
        >
          <div>
            DD MMM YYYY
          </div>
          <div>
            <div
              className=""
            />
            <div>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "Order baru layanan truk ",
                  }
                }
              />
              <span>
                No. Pesanan LT-202108030021
              </span>
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": " oleh <b>BUDI JAYA SEJAHTERA ABADI</b>menggunakan truk milik <b>BINTANG EXPRESS</b> dari <b>Kec. Cibeunying Kidul</b> menuju <b>Malang</b> nilai transaksi <b>Rp500.000,00</b>",
                  }
                }
              />
            </div>
          </div>
          <div>
            HH:mm
          </div>
        </div>
      </div>
    </section>
  </div>
</div>
`;
